rust/monovertex/src/forwarder.rs

```diff
@@ -222,85 +224,140 @@ impl Forwarder {
-                    ))
+        // only breaks out of this loop based on the retry strategy unless all the messages have been written to sink
+        // successfully.
+        loop {
+            while attempts <= config().sink_max_retry_attempts {
+                let status = self
+                    .write_to_sink_once(&mut error_map, &mut fallback_msgs, &mut messages_to_send)
                     .await;
+                match status {
+                    Ok(true) => break,
+                    Ok(false) => {
+                        attempts += 1;
+                        warn!(
+                            "Retry attempt {} due to retryable error. Errors: {:?}",
+                            attempts, error_map
+                        );
+                    }
+                    Err(e) => Err(e)?,
                 }
-                Err(e) => return Err(e),
             }
-        }
 
-        if !messages_to_send.is_empty() {
-            return Err(Error::SinkError(format!(
-                "Failed to sink messages after {} attempts. Errors: {:?}",
-                attempts, error_map
-            )));
+            // If after the retries we still have messages to process
+            if !messages_to_send.is_empty() {
+                // check what is the failure strategy in the config
+                let strategy = config().sink_retry_on_fail_strategy.clone();
+                match strategy.as_str() {
+                    "retry" => {
+                        // reset the attempts and error_map
+                        attempts = 0;
+                        error_map.clear();
+                        continue;
+                    }
+                    "drop" => {
+                        // log that we are dropping the messages as requested
+                        warn!(
+                            "Dropping messages after {} attempts. Errors: {:?}",
+                            attempts, error_map
+                        );
+                        // update the metrics
+                        forward_metrics()
+                            .dropped_total
+                            .get_or_create(&self.common_labels)
+                            .inc_by(messages_to_send.len() as u64);
+                    }
+                    "fallback" => {
+                        fallback_msgs.extend(messages_to_send);
+                    }
+                    _ => {
+                        return Err(Error::SinkError(format!(
+                            "Invalid sink retry on fail strategy: {}",
+                            strategy
+                        )));
+                    }
+                }
+            }
+            break;
         }
 
+        let fallback_msgs_count = fallback_msgs.len() as u64;
         // If there are fallback messages, write them to the fallback sink
-        if !fallback_msgs.is_empty() {
+        if fallback_msgs_count > 0 {
```

Comment:
why did you change this?

Reply:
`let fallback_msgs_count = fallback_msgs.len() as u64;`
I had extracted the count because I had to use it for metrics, have updated there. So can revert this now